### PR TITLE
Audit Apis

### PR DESCRIPTION
**Pull request by TanyaSravan**
## Summary by CodeRabbit

* **New Features**
  * New audit REST APIs to fetch audit entries with filters (project, resource, type, action, user-type, after-timestamp) and pagination; global admin query across projects.
  * Configurable user-type prefix mappings and new DB indexes to speed audit queries.

* **Tests**
  * Comprehensive tests for filters (including user-type and after-timestamp), pagination, limits, defaults, and edge cases.

* **Chores**
  * Updated Docker API version in CI build.

### CODE REVIEW

File: server/src/main/java/com/flipkart/grayskull/controllers/AuditController.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+     * @param offset Pagination offset (default: 0)
+     * @param limit Pagination limit (default: 10, max: 100)
+     * @return ResponseTemplate containing list of audit entries
+     */
+    @Operation(summary = "Retrieves audit entries for a project with optional filtering by resource name, type, and action")
+    @GetMapping("/projects/{projectId}")
+    @PreAuthorize("@grayskullSecurity.hasPermission(#projectId, 'audit.read')")
+    public ResponseTemplate<AuditEntriesResponse> getProjectAudits(
+            @PathVariable("projectId") @Size(max = 255) String projectId,
+            @RequestParam(name = "resourceName", required = false) @Size(max = 500) String resourceName,
```
Comment (grssam) [inline server/src/main/java/com/flipkart/grayskull/controllers/AuditController.java:56]: are we planning to support anything other than "secrets" here in resource? why the generic naming?

File: server/src/main/java/com/flipkart/grayskull/service/interfaces/AuditService.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+     * 
+     * @param projectId Optional project ID filter
+     * @param resourceName Optional resource name filter
+     * @param resourceType Optional resource type filter
+     * @param action Optional action filter
+     * @param offset Pagination offset
+     * @param limit Pagination limit
+     * @return AuditEntriesResponse containing filtered audit entries
+     */
+    AuditEntriesResponse getAuditEntries(String projectId, String resourceName, String resourceType, String action, int offset, int limit);
```
Comment (grssam) [inline server/src/main/java/com/flipkart/grayskull/service/interfaces/AuditService.java]: let's define optional things with Optional

File: server/src/main/java/com/flipkart/grayskull/spimpl/repositories/mongo/AuditEntryMongoRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+public class AuditEntryMongoRepositoryImpl {
+
+    private final MongoTemplate mongoTemplate;
+
+    public List<AuditEntryEntity> findByFilters(String projectId, String resourceName, String resourceType, String action, int offset, int limit) {
+
+        Query query = buildFilterQuery(projectId, resourceName, resourceType, action);
+        query.with(Sort.by(Sort.Direction.DESC, "timestamp"));
+        query.skip(offset);
+        query.limit(limit);
```
Comment (grssam) [inline server/src/main/java/com/flipkart/grayskull/spimpl/repositories/mongo/AuditEntryMongoRepositoryImpl.java]: how heavy is this operation? i.e. sorting by timestamp...

is there any way to optimize this?
we need to keep in mind that this also includes read audits which might be a lot and sorting is happening irrespective of offset/limit..

for a project, across all secrets, for 2 months, it can easily reach 10K+ entries. Let's do some basic test to see if its still optimal and doesn't timeout.

File: server/src/main/java/com/flipkart/grayskull/controllers/AuditController.java (hunk 1 above)
Comment (TanyaSravan) [reply server/src/main/java/com/flipkart/grayskull/controllers/AuditController.java:56]: I was following the naming convention we use in our audits, but we will be supporting policies also right as a resource in the future?

File: server/src/main/java/com/flipkart/grayskull/spimpl/repositories/mongo/AuditEntryMongoRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+/**
+ * MongoDB-specific implementation of custom AuditEntry queries.
+ * Uses MongoTemplate for dynamic query building.
+ * 
+ * Note: Spring Data automatically detects this as the implementation for
+ * AuditEntryMongoRepository's custom methods based on naming convention (*Impl).
+ */
+@RequiredArgsConstructor
+public class AuditEntryMongoRepositoryImpl {
```
Comment (ssrangisetti) [inline server/src/main/java/com/flipkart/grayskull/spimpl/repositories/mongo/AuditEntryMongoRepositoryImpl.java]: how do we ensure that methods in interface and this are matching? is it possible to make this implement audit mongo repository

File: server/src/main/java/com/flipkart/grayskull/spimpl/repositories/mongo/AuditEntryMongoRepositoryImpl.java (hunk 3 above)
Comment (ssrangisetti) [reply server/src/main/java/com/flipkart/grayskull/spimpl/repositories/mongo/AuditEntryMongoRepositoryImpl.java]: instead of sorting all entries, we can take one more filter which gives entries after a certain time. that way we can take advantage of indexes

File: server/src/main/java/com/flipkart/grayskull/controllers/AuditController.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            @RequestParam(name = "resourceType", required = false) @Size(max = 100) String resourceType,
+            @RequestParam(name = "action", required = false) AuditAction action,
+            @RequestParam(name = "userType", required = false) UserType userType,
+            @RequestParam(name = "afterTimestamp", required = false) String afterTimestamp,
+            @RequestParam(name = "offset", defaultValue = "0") @Min(0) int offset,
+            @RequestParam(name = "limit", defaultValue = "10") @Min(1) @Max(100) int limit) {
+        
+        AuditEntriesResponse response = auditService.getAuditEntries(Optional.of(projectId), Optional.ofNullable(resourceName), Optional.ofNullable(resourceType), Optional.ofNullable(action), Optional.ofNullable(userType), parseTimestamp(afterTimestamp), offset, limit);
```
Comment (grssam) [inline server/src/main/java/com/flipkart/grayskull/controllers/AuditController.java]: projectIId is required field. shouldn't be optional

**Comment by grssam** [comment]
Please work on increasing coverage.

**Review by ssrangisetti — APPROVED** [review]

**Review by grssam — APPROVED** [review]
